sessiontxn/staleread/provider.go

```diff
@@ -109,3 +110,28 @@ func (p *StalenessTxnContextProvider) AdviseWarmup() error {
 func (p *StalenessTxnContextProvider) AdviseOptimizeWithPlan(_ interface{}) error {
 	return nil
 }
+
+// GetReadSnapshot get snapshot with read ts
+func (p *StalenessTxnContextProvider) GetReadSnapshot() (kv.Snapshot, error) {
+	return p.getStalenessSnapshot()
+}
+
+// GetForUpdateSnapshot get snapshot with for update ts
+func (p *StalenessTxnContextProvider) GetForUpdateSnapshot() (kv.Snapshot, error) {
+	return p.getStalenessSnapshot()
+}
+
+func (p *StalenessTxnContextProvider) getStalenessSnapshot() (kv.Snapshot, error) {
+	snapshotTS := p.ts
+	txn, err := p.sctx.Txn(false)
+	if err != nil {
+		return nil, err
+	}
+
+	txnCtx := p.sctx.GetSessionVars().TxnCtx
+	if txn.Valid() && txnCtx.StartTS == txnCtx.GetForUpdateTS() && txnCtx.StartTS == snapshotTS {
```

Comment:
We can directly return `txn.GetSnapshot()` if `txn.Valid()` because `txnCtx.StartTS == txnCtx.GetForUpdateTS() && txnCtx.StartTS == snapshotTS` will always be true

Reply:
Done.